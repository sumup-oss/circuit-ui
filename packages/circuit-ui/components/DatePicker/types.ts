/**
 * Copyright 2023, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { WcDatepickerCustomEvent } from 'wc-datepicker';
import type { MonthChangedEventDetails } from 'wc-datepicker/dist/types/components/wc-datepicker/wc-datepicker';

export type MonthChangeEvent =
  WcDatepickerCustomEvent<MonthChangedEventDetails>;

export type DateChangeEvent = WcDatepickerCustomEvent<string>;
export type DateRangeChangeEvent = WcDatepickerCustomEvent<[string, string]>;

export interface DateProps {
  range?: false;
  value?: string;
  onChange?: (event: DateChangeEvent) => void;
}
export interface DateRangeProps {
  range?: true;
  value?: string[];
  onChange?: (event: DateRangeChangeEvent) => void;
}

export interface BaseCalendarProps {
  disabled?: boolean;
  disableDate?: (date: Date) => boolean;
  locale?: string;
}
