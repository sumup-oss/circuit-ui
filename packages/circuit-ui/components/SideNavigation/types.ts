/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FC, MouseEvent, KeyboardEvent, AnchorHTMLAttributes } from 'react';
import { IconProps } from '@sumup/icons';

import { TrackingProps } from '../../hooks/useClickEvent';
import { BadgeProps } from '../Badge';

export interface PrimaryLinkProps
  extends AnchorHTMLAttributes<HTMLAnchorElement> {
  /**
   * Display an icon in addition to the text to help to identify the link.
   * On narrow viewports, only the icon is displayed.
   */
  icon: FC<IconProps>;
  /**
   * Short label to describe the target of the link.
   */
  label: string;
  /**
   * Function that's called when the link is clicked.
   */
  onClick?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * Whether the link is the currently active page.
   */
  isActive?: boolean;
  /**
   * Whether the link is the currently active page.
   */
  isExternal?: boolean;
  /**
   * Whether to show a small circular badge to indicate that a nested secondary
   * link has a badge.
   */
  badge?: boolean;
  /**
   * A collection of secondary groups with nested secondary navigation links.
   */
  secondaryGroups?: SecondaryGroupProps[];
  /**
   * @deprecated
   *
   * Use an `onClick` handler to dispatch user interaction events instead.
   */
  tracking?: TrackingProps;
}

export interface SecondaryGroupProps {
  /**
   * A label that is displayed above the secondary navigation. Only optional
   * for the first group.
   */
  label?: string;
  /**
   * A collection of secondary navigation links.
   */
  secondaryLinks: SecondaryLinkProps[];
  /**
   * @deprecated
   *
   * An optional label that is added to the element tree when clicking a nested
   * secondary navigation link.
   */
  trackingLabel?: string;
}

export interface SecondaryLinkProps {
  /**
   * Short label to describe the target of the link.
   */
  label: string;
  /**
   * A valid path or URL to the link target.
   */
  href: string;
  /**
   * Function that's called when the link is clicked.
   */
  onClick?: (event: MouseEvent | KeyboardEvent) => void;
  /**
   * Whether the link is the currently active page.
   */
  isActive?: boolean;
  /**
   * @deprecated
   *
   * Use an `onClick` handler to dispatch user interaction events instead.
   */
  tracking?: TrackingProps;
  /**
   * An optional badge to highlight the secondary link, e.g. to promote
   * a new link or to indicate new content.
   */
  badge?: BadgeProps;
}
