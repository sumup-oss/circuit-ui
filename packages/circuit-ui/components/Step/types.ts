/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { TrackingProps } from '../../hooks/useClickEvent';
import { ClickEvent } from '../../types/events';

export type Play = () => void;
export type Pause = () => void;
export type Next = () => void;
export type Previous = () => void;

export type Duration = number | ((step?: number) => number);

export type InternalState = {
  step: number;
  previousStep: number;
  paused: boolean;
};

export type State = InternalState & {
  stepDuration: number;
  animationDuration: number;
};

export type Actions = {
  play: Play;
  pause: Pause;
  next: Next;
  previous: Previous;
};

type BaseProps = { onClick?: () => void; [key: string]: unknown };
type ReturnProps<Props> = Props & {
  'aria-label': string;
  'onClick'?: (event: ClickEvent) => void;
};

type PropGetter = <Props extends BaseProps = BaseProps>(
  props?: Props,
) => ReturnProps<Props>;

export type PropGetters = {
  getPlayControlProps: PropGetter;
  getPauseControlProps: PropGetter;
  getNextControlProps: PropGetter;
  getPreviousControlProps: PropGetter;
};

export type StateAndHelpers = PropGetters & {
  state: State;
  actions: Actions;
};
export interface StepOptions {
  /**
   * The total number of steps. Defaults to `0`.
   */
  totalSteps?: number;
  /**
   * The initial step of component. Defaults to `0`.
   */
  initialStep?: number;
  /**
   * Indicates if component should restart after last step. Defaults to `false`.
   */
  cycle?: boolean;
  /**
   * Enable immediate playing of a component after load.
   * Requires `stepDuration` to have any effect. Defaults to `false`.
   */
  autoPlay?: boolean;
  /**
   * The number of steps to interate when navigating. Defaults to `1`.
   */
  stepInterval?: number;
  /**
   * Indicates how long each step will stay visible (in milliseconds).
   * This prop is required for componenets play/pause functionality.
   * If used as function receives state as argument which allows to configure
   * duration based on the data per step. Function should return a number.
   * Defaults to 0.
   */
  stepDuration?: Duration;
  /**
   * Indicates duration of animation between steps (in milliseconds).
   * If used as function receives state as argument which allows to configure
   * duration based on the data per step. Function should return a number.
   * Defaults to 0.
   */
  animationDuration?: Duration;
  /**
   * Function called when play action is triggered.
   */
  onPlay?: (stateAndHelpers: StateAndHelpers) => void;
  /**
   * Function called when pause action is triggered.
   */
  onPause?: (stateAndHelpers: StateAndHelpers) => void;
  /**
   * Function called when go to next step action is triggered.
   */
  onNext?: (stateAndHelpers: StateAndHelpers) => void;
  /**
   * Function called when go to previous step action is triggered.
   */
  onPrevious?: (stateAndHelpers: StateAndHelpers) => void;
  /**
   * @deprecated
   *
   * Use custom event handlers to dispatch user interaction events instead.
   */
  tracking?: TrackingProps;
}
