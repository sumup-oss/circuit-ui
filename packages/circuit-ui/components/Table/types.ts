/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ReactNode } from 'react';

export type SortByValue = boolean | number | string | Date;

export type Direction = 'ascending' | 'descending';

export type CellAlignment = 'left' | 'right' | 'center';

type SortableHeaderCell =
  | { sortable?: false; sortLabel?: never; sortByValue?: never }
  | {
      /**
       * Makes a table column sortable. An accessible `sortLabel` also needs to
       * be provided.
       */
      sortable: true;
      /**
       * Visually hidden label for the sort button for visually impaired users.
       * When passed as a function, it is called with the sort `{ direction }`.
       */
      sortLabel:
        | string
        | (({ direction }: { direction?: Direction }) => string);
    };

type SortableRowCell = {
  /**
   * An optional value to change the order of the table rows.
   */
  sortByValue?: SortByValue;
};

type CellObject = {
  'children': ReactNode;
  'align'?: CellAlignment;
  'data-testid'?: string;
};

export type RowCellObject = SortableRowCell & CellObject;

export type HeaderCellObject = CellObject & SortableHeaderCell;

type Cell = string | number | null | undefined;

export type HeaderCell = Cell | HeaderCellObject;

export type RowCell = Cell | RowCellObject;

export type Row = RowCell[] | { cells: RowCell[] };

export type SortParams =
  | {
      /**
       * Defines whether or not the row is sortable.
       */
      sortable: true;
      /**
       * Defines whether or not the row is currently sorted.
       */
      isSorted: boolean;
      /**
       * A visually hidden label for the sort button for visually impaired users.
       */
      sortLabel: string;
      /**
       * The row's current sort direction.
       */
      sortDirection?: Direction;
    }
  | {
      sortable: false;
      sortLabel?: never;
      sortDirection?: never;
      isSorted?: never;
    };
