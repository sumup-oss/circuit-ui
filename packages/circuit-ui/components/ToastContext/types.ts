/**
 * Copyright 2021, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ClickEvent } from '../../types/events.js';

type OnClose = (event?: ClickEvent) => void;

export interface BaseToastProps {
  /**
   * Callback function that is called when the Toast is closed.
   */
  onClose?: OnClose;
  /**
   * Toast duration, defaults to 6000ms.
   */
  duration?: number;
}

export type ToastComponent<TProps extends BaseToastProps = BaseToastProps> = ((
  props: TProps,
) => JSX.Element) & { TRANSITION_DURATION: number };
