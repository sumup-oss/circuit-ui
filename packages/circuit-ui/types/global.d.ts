/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Theme as CircuitTheme } from '@sumup/design-tokens';
import {} from '@emotion/react/types/css-prop'; // See https://github.com/emotion-js/emotion/pull/1941

declare module '*.mdx' {
  const MDXComponent: (props: any) => JSX.Element;
  export default MDXComponent;
}

declare module '@emotion/react' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  export interface Theme extends CircuitTheme {}
}
