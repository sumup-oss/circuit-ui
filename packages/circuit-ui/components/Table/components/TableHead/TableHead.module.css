.fixed {
  transform: translateY(var(--table-head-top));
}

@media (max-width: 767px) {
  .fixed {
    transform: translateY(var(--table-head-top));
  }
}

@media (max-width: 767px) {
  .row-headers {
    transform: unset;
  }
}
