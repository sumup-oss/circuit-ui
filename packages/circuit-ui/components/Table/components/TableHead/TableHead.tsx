/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Fragment } from 'react';
import { css } from '@emotion/react';

import TableRow from '../TableRow';
import TableHeader from '../TableHeader';
import { mapCellProps, getSortParams } from '../../utils';
import { Direction, HeaderCell } from '../../types';
import styled, { StyleProps } from '../../../../styles/styled';

type ScrollableOptions =
  | {
      /**
       * Adds scrollable styles the table head according to the table props.
       */
      scrollable: true;
      /**
       * The value used to make the thead fixed while scrolling.
       */
      top: number;
    }
  | {
      scrollable?: boolean;
      top?: number;
    };

type TableHeadProps = ScrollableOptions & {
  /**
   * An array of header cells for the table.
   */
  headers?: HeaderCell[];
  /**
   * Enables/disables sticky columns on mobile
   */
  rowHeaders?: boolean;
  /**
   * Adds condensed styles the table head according to the table props.
   */
  condensed?: boolean;
  /**
   * sortBy handler
   */
  onSortBy?: (i: number) => void;
  /**
   * The current sortDirection
   */
  sortDirection?: Direction;
  /**
   * The current sorted row index
   */
  sortedRow?: number;
  /**
   * sortEnter handler
   */
  onSortEnter?: (i: number) => void;
  /**
   * sortLeave handler
   */
  onSortLeave?: (i: number) => void;
};

type TheadElProps = Pick<TableHeadProps, 'scrollable' | 'top' | 'rowHeaders'>;

const fixedStyles = ({
  scrollable,
  top,
  rowHeaders,
  theme,
}: TheadElProps & StyleProps) =>
  scrollable &&
  top && // we need this check despite the TS types
  css`
    transform: translateY(${top}px);

    ${theme.mq.untilMega} {
      transform: ${rowHeaders ? 'unset' : `translateY(${top}px)`};
    }
  `;

const Thead = styled.thead<TheadElProps>`
  ${fixedStyles}
`;

/**
 * TableHead for the Table component. The Table handles rendering it.
 */
const TableHead = ({
  headers = [],
  rowHeaders = false,
  condensed,
  scrollable,
  top,
  onSortBy,
  sortDirection,
  sortedRow,
  onSortEnter,
  onSortLeave,
}: TableHeadProps): JSX.Element => (
  <Thead scrollable={scrollable} top={top} rowHeaders={rowHeaders}>
    {!!headers.length && (
      <TableRow>
        {headers.map((header, i) => {
          const cellProps = mapCellProps(header);
          const { sortable, sortLabel } = cellProps;
          const sortParams = getSortParams({
            rowIndex: i,
            sortable,
            sortDirection,
            sortLabel,
            sortedRow,
          });
          return (
            <Fragment key={`table-header-${i}`}>
              <TableHeader
                {...cellProps}
                condensed={condensed}
                fixed={rowHeaders && i === 0}
                onClick={
                  sortParams.sortable
                    ? onSortBy && (() => onSortBy(i))
                    : undefined
                }
                onMouseEnter={
                  sortParams.sortable
                    ? onSortEnter && (() => onSortEnter(i))
                    : undefined
                }
                onMouseLeave={
                  sortParams.sortable
                    ? onSortLeave && (() => onSortLeave(i))
                    : undefined
                }
                sortParams={sortParams}
              />
            </Fragment>
          );
        })}
      </TableRow>
    )}
  </Thead>
);

export default TableHead;
