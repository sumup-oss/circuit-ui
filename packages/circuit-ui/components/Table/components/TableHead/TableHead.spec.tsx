/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  create,
  render,
  renderToHtml,
  axe,
  act,
  userEvent,
} from '../../../../util/test-utils';
import { Cell, Direction } from '../../types';

import TableHead from '.';

const sortLabel = ({ direction }: { direction?: Direction }) => {
  const order = direction === 'ascending' ? 'descending' : 'ascending';
  return `Sort in ${order} order`;
};
const fixtureHeaders: Cell[] = [
  { children: 'Foo', sortable: true, sortLabel },
  'Bar',
  'Baz',
];

describe('TableHead', () => {
  describe('Style tests', () => {
    it('should render with default styles', () => {
      const actual = create(<TableHead headers={fixtureHeaders} />);
      expect(actual).toMatchSnapshot();
    });

    it('should render with rowHeader styles', () => {
      const actual = create(<TableHead rowHeaders headers={fixtureHeaders} />);
      expect(actual).toMatchSnapshot();
    });
  });

  describe('onClick', () => {
    it('should not dispatch the onSortBy handler when the column is not sortable', () => {
      const headers = ['Foo'];
      const onSortByMock = jest.fn();
      const { getByRole } = render(
        <TableHead onSortBy={onSortByMock} headers={headers} />,
      );

      act(() => {
        userEvent.click(getByRole('columnheader'));
      });

      expect(onSortByMock).not.toHaveBeenCalled();
    });

    it('should dispatch the onSortBy handler', () => {
      const headers: Cell[] = [{ children: 'Foo', sortable: true, sortLabel }];
      const onSortByMock = jest.fn();
      const { getByRole } = render(
        <TableHead onSortBy={onSortByMock} headers={headers} />,
      );

      act(() => {
        userEvent.click(getByRole('columnheader'));
      });

      expect(onSortByMock).toHaveBeenCalledTimes(1);
      expect(onSortByMock).toHaveBeenCalledWith(0);
    });
  });

  describe('onSortEnter', () => {
    it('should not dispatch the onSortEnter handler when the column is not sortable', () => {
      const headers = ['Foo'];
      const onSortEnterMock = jest.fn();
      const { getByRole } = render(
        <TableHead onSortEnter={onSortEnterMock} headers={headers} />,
      );

      act(() => {
        userEvent.hover(getByRole('columnheader'));
      });

      expect(onSortEnterMock).not.toHaveBeenCalled();
    });

    it('should dispatch the onSortEnter handler', () => {
      const headers: Cell[] = [{ children: 'Foo', sortable: true, sortLabel }];
      const onSortEnterMock = jest.fn();
      const { getByRole } = render(
        <TableHead onSortEnter={onSortEnterMock} headers={headers} />,
      );

      act(() => {
        userEvent.hover(getByRole('columnheader'));
      });

      expect(onSortEnterMock).toHaveBeenCalledTimes(1);
      expect(onSortEnterMock).toHaveBeenCalledWith(0);
    });
  });

  describe('onSortLeave', () => {
    it('should dispatch the onSortLeave handler', () => {
      const headers: Cell[] = [{ children: 'Foo', sortable: true, sortLabel }];
      const onSortLeaveMock = jest.fn();
      const { getByRole } = render(
        <TableHead onSortLeave={onSortLeaveMock} headers={headers} />,
      );

      act(() => {
        userEvent.unhover(getByRole('columnheader'));
      });

      expect(onSortLeaveMock).toHaveBeenCalledTimes(1);
      expect(onSortLeaveMock).toHaveBeenCalledWith(0);
    });

    it('should not dispatch the onSortLeave handler when the column is not sortable', () => {
      const headers = ['Foo'];
      const onSortLeaveMock = jest.fn();
      const { getByRole } = render(
        <TableHead onSortLeave={onSortLeaveMock} headers={headers} />,
      );

      act(() => {
        userEvent.unhover(getByRole('columnheader'));
      });

      expect(onSortLeaveMock).not.toHaveBeenCalled();
    });
  });

  describe('Accessibility tests', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(
        <TableHead rowHeaders headers={fixtureHeaders} />,
      );
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
