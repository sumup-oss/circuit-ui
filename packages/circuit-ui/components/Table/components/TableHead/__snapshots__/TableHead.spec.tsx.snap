// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHead Style tests should render with default styles 1`] = `
.circuit-1 {
  vertical-align: middle;
}

tbody .circuit-1:last-child th,
tbody .circuit-1:last-child td {
  border-bottom: none;
}

.circuit-3 {
  background-color: var(--cui-bg-normal);
  border-bottom: 1px solid var(--cui-border-divider);
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: var(--cui-fg-subtle);
  font-weight: 700;
  padding: 8px 24px;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.circuit-3:focus-within::after {
  content: '';
  display: block;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  outline: 0;
  box-shadow: 0 0 0 4px var(--cui-border-focus);
}

.circuit-3:focus-within::after::-moz-focus-inner {
  border: 0;
}

.circuit-3:focus-within,
.circuit-3:hover {
  background-color: var(--cui-bg-normal-hovered);
  color: var(--cui-fg-accent-hovered);
}

.circuit-3:focus-within>button,
.circuit-3:hover>button {
  opacity: 1;
}

.circuit-3:active {
  background-color: var(--cui-bg-normal-pressed);
  color: var(--cui-fg-accent-pressed);
}

.circuit-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 36px;
  width: 24px;
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 120ms ease-in-out;
  transition: opacity 120ms ease-in-out;
  color: var(--cui-fg-accent);
  border: 0;
  background: none;
  outline: 0;
  padding: 2px 4px;
  margin: 0;
  cursor: pointer;
}

.circuit-4:focus {
  opacity: 1;
}

.circuit-4:focus::-moz-focus-inner {
  border: 0;
}

.circuit-5 {
  margin: -2px 0;
}

.circuit-7 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-8 {
  background-color: var(--cui-bg-normal);
  border-bottom: 1px solid var(--cui-border-divider);
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: var(--cui-fg-subtle);
  font-weight: 700;
  padding: 8px 24px;
  vertical-align: middle;
  white-space: nowrap;
}

<thead
  class="circuit-0"
>
  <tr
    class="circuit-1 circuit-2"
  >
    <th
      aria-label="Sort in ascending order"
      aria-sort="none"
      class="circuit-3"
      scope="col"
    >
      <button
        class="circuit-4"
        role="button"
        title="Sort in ascending order"
      >
        <svg
          class="circuit-5"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 5c.266 0 .52.104.71.29l4 4c.186.19.29.444.29.71a1 1 0 0 1-1 1c-.265 0-.52-.104-.71-.29L8 7.42l-3.29 3.29c-.19.186-.444.29-.71.29a1 1 0 0 1-1-1c0-.266.104-.52.29-.71l4-4C7.48 5.104 7.734 5 8 5z"
            fill="currentColor"
          />
        </svg>
        <svg
          class="circuit-5"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11c-.266 0-.52-.104-.71-.29l-4-4A1.013 1.013 0 0 1 3 6a1 1 0 0 1 1-1c.266 0 .52.104.71.29L8 8.58l3.29-3.29c.19-.186.445-.29.71-.29a1 1 0 0 1 1 1c0 .266-.104.52-.29.71l-4 4c-.19.186-.444.29-.71.29z"
            fill="currentColor"
          />
        </svg>
        <span
          class="circuit-7"
        >
          Sort in ascending order
        </span>
      </button>
      Foo
    </th>
    <th
      class="circuit-8"
      scope="col"
    >
      Bar
    </th>
    <th
      class="circuit-8"
      scope="col"
    >
      Baz
    </th>
  </tr>
</thead>
`;

exports[`TableHead Style tests should render with rowHeader styles 1`] = `
.circuit-1 {
  vertical-align: middle;
}

tbody .circuit-1:last-child th,
tbody .circuit-1:last-child td {
  border-bottom: none;
}

.circuit-3 {
  background-color: var(--cui-bg-normal);
  border-bottom: 1px solid var(--cui-border-divider);
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: var(--cui-fg-subtle);
  font-weight: 700;
  padding: 8px 24px;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

@media (max-width: 767px) {
  .circuit-3 {
    left: 0;
    top: auto;
    position: absolute;
    width: 145px;
    overflow-wrap: break-word;
  }
}

.circuit-3:focus-within::after {
  content: '';
  display: block;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 1;
  outline: 0;
  box-shadow: 0 0 0 4px var(--cui-border-focus);
}

.circuit-3:focus-within::after::-moz-focus-inner {
  border: 0;
}

.circuit-3:focus-within,
.circuit-3:hover {
  background-color: var(--cui-bg-normal-hovered);
  color: var(--cui-fg-accent-hovered);
}

.circuit-3:focus-within>button,
.circuit-3:hover>button {
  opacity: 1;
}

.circuit-3:active {
  background-color: var(--cui-bg-normal-pressed);
  color: var(--cui-fg-accent-pressed);
}

.circuit-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  height: 36px;
  width: 24px;
  position: absolute;
  left: 0;
  top: 50%;
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -moz-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 120ms ease-in-out;
  transition: opacity 120ms ease-in-out;
  color: var(--cui-fg-accent);
  border: 0;
  background: none;
  outline: 0;
  padding: 2px 4px;
  margin: 0;
  cursor: pointer;
}

.circuit-4:focus {
  opacity: 1;
}

.circuit-4:focus::-moz-focus-inner {
  border: 0;
}

.circuit-5 {
  margin: -2px 0;
}

.circuit-7 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-8 {
  background-color: var(--cui-bg-normal);
  border-bottom: 1px solid var(--cui-border-divider);
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 120ms ease-in-out;
  transition: background-color 120ms ease-in-out;
  vertical-align: middle;
  overflow-wrap: break-word;
  display: none;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 700;
  padding: 8px 24px;
  white-space: nowrap;
}

@media (max-width: 767px) {
  .circuit-8 {
    display: table-cell;
    min-width: 145px;
    max-width: 145px;
    width: 145px;
  }
}

.circuit-9 {
  background-color: var(--cui-bg-normal);
  border-bottom: 1px solid var(--cui-border-divider);
  padding: 24px;
  text-align: left;
  -webkit-transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  transition: background-color 120ms ease-in-out,color 120ms ease-in-out;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: var(--cui-fg-subtle);
  font-weight: 700;
  padding: 8px 24px;
  vertical-align: middle;
  white-space: nowrap;
}

<thead
  class="circuit-0"
>
  <tr
    class="circuit-1 circuit-2"
  >
    <th
      aria-label="Sort in ascending order"
      aria-sort="none"
      class="circuit-3"
      scope="col"
    >
      <button
        class="circuit-4"
        role="button"
        title="Sort in ascending order"
      >
        <svg
          class="circuit-5"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 5c.266 0 .52.104.71.29l4 4c.186.19.29.444.29.71a1 1 0 0 1-1 1c-.265 0-.52-.104-.71-.29L8 7.42l-3.29 3.29c-.19.186-.444.29-.71.29a1 1 0 0 1-1-1c0-.266.104-.52.29-.71l4-4C7.48 5.104 7.734 5 8 5z"
            fill="currentColor"
          />
        </svg>
        <svg
          class="circuit-5"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 11c-.266 0-.52-.104-.71-.29l-4-4A1.013 1.013 0 0 1 3 6a1 1 0 0 1 1-1c.266 0 .52.104.71.29L8 8.58l3.29-3.29c.19-.186.445-.29.71-.29a1 1 0 0 1 1 1c0 .266-.104.52-.29.71l-4 4c-.19.186-.444.29-.71.29z"
            fill="currentColor"
          />
        </svg>
        <span
          class="circuit-7"
        >
          Sort in ascending order
        </span>
      </button>
      Foo
    </th>
    <td
      aria-hidden="true"
      class="circuit-8"
      role="presentation"
    >
      Foo
    </td>
    <th
      class="circuit-9"
      scope="col"
    >
      Bar
    </th>
    <th
      class="circuit-9"
      scope="col"
    >
      Baz
    </th>
  </tr>
</thead>
`;
